lib/redis_memo/memoize_method.rb

```diff
@@ -81,14 +81,61 @@ def self.extract_dependencies(ref, *method_args, &depends_on)
 
   def self.get_or_extract_dependencies(ref, *method_args, &depends_on)
     if RedisMemo::Cache.local_dependency_cache
-      RedisMemo::Cache.local_dependency_cache[ref] ||= {}
-      RedisMemo::Cache.local_dependency_cache[ref][depends_on] ||= {}
-      RedisMemo::Cache.local_dependency_cache[ref][depends_on][method_args] ||= extract_dependencies(ref, *method_args, &depends_on)
+      RedisMemo::Cache.local_dependency_cache[ref.class] ||= {}
+      RedisMemo::Cache.local_dependency_cache[ref.class][depends_on] ||= {}
+      mapped_args = exclude_anonymous_args(depends_on, ref, method_args)
+      RedisMemo::Cache.local_dependency_cache[ref.class][depends_on][mapped_args] ||= extract_dependencies(ref, *method_args, &depends_on)
     else
       extract_dependencies(ref, *method_args, &depends_on)
     end
   end
 
+  # We only look at known method parameters in the dependency block in order to define its dependent
+  # memos and ignore anonymous parameters, following the convention that nil or :_ is an anonymous parameter.
+  # Example:
+  # ```
+  #    def method(param1, param2)
+  #    end
+  #
+  #    memoize_method :method do |_, _, param2|`
+  #      depends_on RedisMemo::Memoizable.new(param2: param2)
+  #    end
+  # ```
+  #  `exclude_anonymous_args(depends_on, ref, [1, 2])` returns { param2 : 2 }
+  def self.exclude_anonymous_args(depends_on, ref, args)
+    mapped_args = {}
+    unless depends_on.parameters.empty? or args.empty?
+      depends_on_args = [ref] + args
+      j = 0
+
+      depends_on.parameters.each_with_index do |param, i|
+        unless param.size != 2 || param[1] == :_
+          case param[0]
+          # If the parameter is a splat, we take the rest of the arguments if it's the last parameter. Otherwise,
+          # parameters that come after a splat can only be keyword args / hashes.
+          when :rest
+            if i == depends_on.parameters.size - 1
+              mapped_args[param[1]] = depends_on_args[j..-1]
+            else
+              single_splat_args = depends_on_args[j..-1].select { |x| !x.is_a?(Hash) }
```

Comment:
The nested loop is still here via this select.

Want to continue the discussion at https://github.com/chanzuckerberg/redis-memo/pull/40#discussion_r578666164

Do we really need to map each arg to its named parameter tho?


```ruby
memoize_method :test do |_, a, *args,  b:, **kwargs| end

test(1, 2, 3, 4, b: 5, c: 6)
```
if we try to `exclude_anonymous_args`, the answer shoud just be `[1, 2, 3, 4, {b: 5, c: 6}]` right?

Reply:
Basically, first figure out when do we need to splat, then loop through `parameters` again and ignore the anonymous_args.

Reply:
Yea you're right, we don't really need to map it to the named parameter.

But we could have something like 
```ruby
memoize_method :test do |_, *args, b:, **|

test(1, 2, 3, 4, b: 5, c: 6, d: 6)
```
In which case, the answer would be 
`[2,3,4, {b: 5}]`
We can't just look at/discard the last element of the method_args, since it would be discarding `{b: 5, c: 6, d: 6}`. So I'm not sure how to avoid looping through the rest of the array with `select`?